File: frontend/src/lib/api/worker.ts
import { API_BASE_URL } from "./client";
import { WorkerIssueReport, WorkerIssueOut } from "../types/domain";
import { offlineDb } from "../db/offline-db";

/**
 * Submit an on-ground worker grievance / hazard issue.
 * Supports transparent offline storage via IndexedDB Dexie.
 */
export async function reportWorkerIssue(payload: WorkerIssueReport): Promise<WorkerIssueOut> {
  const isOnline = typeof navigator !== "undefined" ? navigator.onLine : true;

  if (isOnline) {
    try {
      const res = await fetch(`${API_BASE_URL}/worker/report-issue`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(payload),
      });
      if (res.ok) {
        const data: WorkerIssueOut = await res.json();
        // Also save to offlineDb for offline review
        try {
          await offlineDb.worker_issues.put({
            id: data.id,
            mine_site_id: data.mine_site_id,
            issue_category: data.issue_category,
            location_description: data.location_description,
            description: data.description,
            urgency: data.urgency,
            photo_evidence_url: data.photo_evidence_url,
            status: data.status,
            is_synced: 1,
            created_at: data.created_at,
          });
        } catch {}
        return data;
      }
    } catch (err) {
      console.warn("Network request failed, queueing worker issue offline:", err);
    }
  }

  // Offline fallback: store locally
  const offlineId = `offline-${Date.now()}-${Math.random().toString(36).substring(2, 7)}`;
  const now = new Date().toISOString();
  const offlineIssue: WorkerIssueOut = {
    id: offlineId,
    mine_site_id: payload.mine_site_id,
    issue_category: payload.issue_category,
    location_description: payload.location_description,
    description: payload.description,
    urgency: payload.urgency,
    photo_evidence_url: payload.photo_evidence_url,
    status: "PENDING_REVIEW",
    reported_by_name: "Logged Offline (Mining Sirdar)",
    created_at: now,
    is_emergency_stop: payload.urgency === "EMERGENCY_STOP",
  };

  try {
    await offlineDb.worker_issues.put({
      id: offlineId,
      mine_site_id: payload.mine_site_id,
      issue_category: payload.issue_category,
      location_description: payload.location_description,
      description: payload.description,
      urgency: payload.urgency,
      photo_evidence_url: payload.photo_evidence_url,
      status: "PENDING_REVIEW",
      is_synced: 0,
      created_at: now,
    });
  } catch (dbErr) {
    console.warn("Could not save to Dexie, fallback to localStorage:", dbErr);
    try {
      const existing = JSON.parse(localStorage.getItem("offline_worker_issues") || "[]");
      existing.unshift(offlineIssue);
      localStorage.setItem("offline_worker_issues", JSON.stringify(existing));
    } catch {}
  }

  return offlineIssue;
}

/**
 * Fetch reported issues for the current worker / mine site.
 */
export async function fetchMyWorkerIssues(mineSiteId?: string): Promise<WorkerIssueOut[]> {
  try {
    const url = mineSiteId
      ? `${API_BASE_URL}/worker/my-issues?mine_site_id=${encodeURIComponent(mineSiteId)}`
      : `${API_BASE_URL}/worker/my-issues`;

    const res = await fetch(url);
    if (res.ok) {
      const serverIssues: WorkerIssueOut[] = await res.json();
      
      // Merge with any un-synced offline issues
      try {
        const localIssues = await offlineDb.worker_issues.where("is_synced").equals(0).toArray();
        if (localIssues.length > 0) {
          const mappedLocal: WorkerIssueOut[] = localIssues.map((li) => ({
            id: li.id,
            mine_site_id: li.mine_site_id,
            issue_category: li.issue_category,
            location_description: li.location_description,
            description: li.description,
            urgency: li.urgency,
            photo_evidence_url: li.photo_evidence_url,
            status: "PENDING_REVIEW",
            reported_by_name: "Offline Local Queue",
            created_at: li.created_at,
            is_emergency_stop: li.urgency === "EMERGENCY_STOP",
          }));
          return [...mappedLocal, ...serverIssues];
        }
      } catch {}

      return serverIssues;
    }
  } catch (err) {
    console.warn("Failed to fetch worker issues from server, reading local cache:", err);
  }

  // Fallback to local Dexie issues
  try {
    const localIssues = await offlineDb.worker_issues.toArray();
    if (localIssues.length > 0) {
      return localIssues.map((li) => ({
        id: li.id,
        mine_site_id: li.mine_site_id,
        issue_category: li.issue_category,
        location_description: li.location_description,
        description: li.description,
        urgency: li.urgency,
        photo_evidence_url: li.photo_evidence_url,
        status: (li.status as any) || "PENDING_REVIEW",
        reported_by_name: "Local Cache (Offline)",
        created_at: li.created_at,
        is_emergency_stop: li.urgency === "EMERGENCY_STOP",
      }));
    }
  } catch {}

  // Sample fallback
  return [
    {
      id: "demo-issue-01",
      mine_site_id: mineSiteId || "11111111-1111-4111-a111-111111111111",
      issue_category: "VENTILATION_GAS",
      location_description: "Level 3 Gallery 4, 380m near Return Airway",
      description: "Methane sensor flashing intermittent warning; airflow velocity dropped below 0.8 m/s during shift.",
      urgency: "HIGH",
      status: "INVESTIGATING",
      reported_by_name: "K. Shankaraiah (Mining Sirdar)",
      created_at: new Date(Date.now() - 3600000).toISOString(),
      is_emergency_stop: false,
    },
    {
      id: "demo-issue-02",
      mine_site_id: mineSiteId || "11111111-1111-4111-a111-111111111111",
      issue_category: "ROOF_SUPPORT",
      location_description: "Panel 7 Working Face Junction 2",
      description: "Loose shale spalling detected on left rib; SSR roof bolt anchor plate requires retightening.",
      urgency: "EMERGENCY_STOP",
      status: "PENDING_REVIEW",
      reported_by_name: "Rajesh Kumar Mandal (Overman)",
      created_at: new Date(Date.now() - 7200000).toISOString(),
      is_emergency_stop: true,
    },
    {
      id: "demo-issue-03",
      mine_site_id: mineSiteId || "11111111-1111-4111-a111-111111111111",
      issue_category: "EQUIPMENT_DEFECT",
      location_description: "Haulage Incline No. 1 Winch Room",
      description: "Winch brake shoe lining worn out; emergency trip switch tested and lubricated.",
      urgency: "MEDIUM",
      status: "RECTIFIED",
      reported_by_name: "Dilip Hembram",
      created_at: new Date(Date.now() - 86400000).toISOString(),
      is_emergency_stop: false,
    },
  ];
}

import { WorkerLeave, WorkerLeaveRequest } from "../types/domain";

function getAuthHeaders(): HeadersInit {
  if (typeof window === "undefined") return { "Content-Type": "application/json" };
  const authStorage = localStorage.getItem("auth-storage");
  if (authStorage) {
    try {
      const parsed = JSON.parse(authStorage);
      const token = parsed?.state?.accessToken;
      if (token) {
        return {
          "Content-Type": "application/json",
          Authorization: `Bearer ${token}`,
        };
      }
    } catch (e) {
      console.error("Failed to parse auth token", e);
    }
  }
  return { "Content-Type": "application/json" };
}

/**
 * Submit a statutory leave application (Mines Rules 1955 Chapter VII).
 * Supports transparent offline storage via IndexedDB Dexie.
 */
export async function submitWorkerLeave(payload: WorkerLeaveRequest): Promise<WorkerLeave> {
  const isOnline = typeof navigator !== "undefined" ? navigator.onLine : true;

  if (isOnline) {
    try {
      const res = await fetch(`${API_BASE_URL}/worker/leave-applications`, {
        method: "POST",
        headers: getAuthHeaders(),
        body: JSON.stringify(payload),
      });

      if (res.ok) {
        const data: WorkerLeave = await res.json();
        try {
          await offlineDb.worker_leaves.put({
            id: data.id,
            mine_site_id: data.mine_site_id,
            worker_id: data.worker_id,
            worker_name: data.worker_name,
            leave_type: data.leave_type,
            start_date: data.start_date,
            end_date: data.end_date,
            total_days: data.total_days,
            reason: data.reason,
            relief_worker_id: data.relief_worker_id || undefined,
            relief_worker_name: data.relief_worker_name || undefined,
            status: data.status,
            is_synced: 1,
            created_at: data.created_at,
          });
        } catch {}
        return data;
      }
    } catch (err) {
      console.warn("Network request failed, queueing worker leave offline:", err);
    }
  }

  // Offline fallback
  const offlineId = `offline-leave-${Date.now()}`;
  const now = new Date().toISOString();
  const offlineLeave: WorkerLeave = {
    id: offlineId,
    worker_id: payload.worker_id || "W-104",
    worker_name: payload.worker_name || "K. Shankaraiah (Worker)",
    mine_site_id: payload.mine_site_id,
    leave_type: payload.leave_type,
    start_date: payload.start_date,
    end_date: payload.end_date,
    total_days: payload.total_days,
    reason: payload.reason,
    relief_worker_id: payload.relief_worker_id,
    relief_worker_name: payload.relief_worker_name,
    status: "QUEUED_OFFLINE",
    created_at: now,
  };

  try {
    await offlineDb.worker_leaves.put({
      id: offlineId,
      mine_site_id: payload.mine_site_id,
      worker_id: offlineLeave.worker_id,
      worker_name: offlineLeave.worker_name,
      leave_type: payload.leave_type,
      start_date: payload.start_date,
      end_date: payload.end_date,
      total_days: payload.total_days,
      reason: payload.reason,
      relief_worker_id: payload.relief_worker_id || undefined,
      relief_worker_name: payload.relief_worker_name || undefined,
      status: "QUEUED_OFFLINE",
      is_synced: 0,
      created_at: now,
    });
  } catch (dexErr) {
    console.warn("Could not save worker leave to Dexie:", dexErr);
  }

  return offlineLeave;
}

/**
 * Fetch statutory leave applications for the mine site / worker.
 */
export async function fetchWorkerLeaves(mineSiteId?: string | null, workerId?: string | null): Promise<WorkerLeave[]> {
  try {
    const params = new URLSearchParams();
    if (mineSiteId) params.append("mine_site_id", mineSiteId);
    if (workerId) params.append("worker_id", workerId);

    const url = params.toString()
      ? `${API_BASE_URL}/worker/leave-applications?${params.toString()}`
      : `${API_BASE_URL}/worker/leave-applications`;

    const res = await fetch(url, { headers: getAuthHeaders() });
    if (res.ok) {
      const serverLeaves: WorkerLeave[] = await res.json();
      // Merge with unsynced local offline leaves
      try {
        const localLeaves = await offlineDb.worker_leaves.where("is_synced").equals(0).toArray();
        if (localLeaves.length > 0) {
          const mappedLocal: WorkerLeave[] = localLeaves.map((l) => ({
            id: l.id,
            worker_id: l.worker_id,
            worker_name: l.worker_name,
            mine_site_id: l.mine_site_id,
            leave_type: l.leave_type,
            start_date: l.start_date,
            end_date: l.end_date,
            total_days: l.total_days,
            reason: l.reason,
            relief_worker_id: l.relief_worker_id || null,
            relief_worker_name: l.relief_worker_name || null,
            status: l.status,
            created_at: l.created_at,
          }));
          return [...mappedLocal, ...serverLeaves];
        }
      } catch {}

      return serverLeaves;
    }
  } catch (err) {
    console.warn("Failed to fetch worker leaves from server, reading local cache:", err);
  }

  // Fallback to local Dexie leaves
  try {
    const localLeaves = await offlineDb.worker_leaves.toArray();
    if (localLeaves.length > 0) {
      return localLeaves.map((l) => ({
        id: l.id,
        worker_id: l.worker_id,
        worker_name: l.worker_name,
        mine_site_id: l.mine_site_id,
        leave_type: l.leave_type,
        start_date: l.start_date,
        end_date: l.end_date,
        total_days: l.total_days,
        reason: l.reason,
        relief_worker_id: l.relief_worker_id || null,
        relief_worker_name: l.relief_worker_name || null,
        status: l.status,
        created_at: l.created_at,
      }));
    }
  } catch {}

  return [
    {
      id: "leave-demo-1",
      worker_id: "W-104",
      worker_name: "Rajesh Kumar Mandal",
      mine_site_id: mineSiteId || "11111111-1111-4111-a111-111111111111",
      leave_type: "EARNED_STATUTORY",
      start_date: "2026-09-12",
      end_date: "2026-09-14",
      total_days: 3,
      reason: "Annual statutory leave entitlement under Mines Rules 1955 Chapter VII.",
      relief_worker_id: "W-108",
      relief_worker_name: "K. Shankaraiah (Mining Sirdar)",
      status: "APPROVED",
      reviewed_by: "Colliery Manager (Er. Ramesh Rao)",
      reviewed_at: new Date().toISOString(),
      review_notes: "Shift relief allocated to Sirdar Shankaraiah.",
      created_at: new Date().toISOString(),
    },
    {
      id: "leave-demo-2",
      worker_id: "W-104",
      worker_name: "Rajesh Kumar Mandal",
      mine_site_id: mineSiteId || "11111111-1111-4111-a111-111111111111",
      leave_type: "CASUAL",
      start_date: "2026-09-20",
      end_date: "2026-09-21",
      total_days: 2,
      reason: "Family medical appointment in Hyderabad.",
      relief_worker_id: "W-112",
      relief_worker_name: "G. Venkatesh",
      status: "SUBMITTED",
      created_at: new Date().toISOString(),
    },
  ];
}

/**
 * Colliery Manager reviews (Approve/Reject) a worker leave application.
 */
export async function reviewWorkerLeave(
  leaveId: string,
  status: "APPROVED" | "REJECTED",
  reviewNotes?: string
): Promise<WorkerLeave> {
  const res = await fetch(`${API_BASE_URL}/worker/leave-applications/${leaveId}/review`, {
    method: "POST",
    headers: getAuthHeaders(),
    body: JSON.stringify({
      status,
      review_notes: reviewNotes,
    }),
  });

  if (!res.ok) {
    const err = await res.json().catch(() => ({ detail: "Review action failed" }));
    throw new Error(err.detail || "Failed to update leave status");
  }

  return res.json();
}

File: app/api/v1/worker.py
import uuid
from datetime import datetime, date, timezone
from typing import Any, Dict, List, Optional
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel, Field

from app.dependencies import CurrentUser, get_notif_repo, get_capa_repo
from app.domain.enums import UserRole, CAPAState
from app.infrastructure.database.models import (
    NotificationModel,
    ViolationCAPAModel,
    StatutoryRuleModel,
    MineSiteModel,
    WorkerLeaveModel,
)
from app.infrastructure.repositories.notification_repository import NotificationRepository
from app.infrastructure.repositories.capa_repository import CAPARepository

router = APIRouter(prefix="/worker", tags=["Worker Issue & Hazard Portal"])


class WorkerIssueReport(BaseModel):
    mine_site_id: str = Field(..., description="UUID or identifier of the mine site")
    issue_category: str = Field(..., description="e.g. VENTILATION_GAS, ROOF_SUPPORT, EQUIPMENT_DEFECT, WATER_LOGGING, WELFARE")
    location_description: str = Field(..., description="e.g. Gallery 4, 380m Level near Shaft Bottom")
    description: str = Field(..., description="Detailed description of the hazard or grievance")
    urgency: str = Field("MEDIUM", description="LOW, MEDIUM, HIGH, EMERGENCY_STOP")
    photo_evidence_url: Optional[str] = Field(None, description="Optional photo URL or uploaded evidence link")


class WorkerIssueOut(BaseModel):
    id: UUID
    mine_site_id: UUID
    issue_category: str
    location_description: str
    description: str
    urgency: str
    photo_evidence_url: Optional[str] = None
    status: str  # PENDING_REVIEW, INVESTIGATING, RECTIFIED
    reported_by_id: Optional[UUID] = None
    reported_by_name: Optional[str] = None
    created_at: datetime
    updated_at: Optional[datetime] = None
    is_emergency_stop: bool = False
    capa_id: Optional[UUID] = None

    class Config:
        from_attributes = True


CATEGORY_RULE_MAP = {
    "VENTILATION_GAS": "CMR-2017-153",
    "ROOF_SUPPORT": "CMR-2017-123",
    "EQUIPMENT_DEFECT": "CMR-2017-130",
    "WATER_LOGGING": "CMR-2017-145",
    "WELFARE": "CMR-2017-RULE29B",
}


@router.post("/report-issue", response_model=WorkerIssueOut, status_code=201)
async def report_worker_issue(
    body: WorkerIssueReport,
    user: CurrentUser,
    notif_repo: NotificationRepository = Depends(get_notif_repo),
    capa_repo: CAPARepository = Depends(get_capa_repo),
):
    """
    Log an on-ground issue, safety hazard, equipment defect, or ventilation concern.
    Creates an unverified ViolationCAPAModel and alerts the Colliery Manager.
    If urgency == EMERGENCY_STOP, triggers a real-time CRITICAL warning.
    """
    # Parse mine_site_id
    try:
        mine_site_uuid = UUID(body.mine_site_id)
    except Exception:
        # Fallback to default Telangana GDK-11A mine
        mine_site_uuid = UUID("11111111-1111-4111-a111-111111111111")

    user_id = UUID(user["user_id"]) if "user_id" in user else uuid.uuid4()
    user_name = user.get("full_name") or user.get("email") or "Mining Sirdar / Worker"

    # Find matching statutory rule or default
    rule_code = CATEGORY_RULE_MAP.get(body.issue_category, "CMR-2017-123")
    rule = await StatutoryRuleModel.find_one(StatutoryRuleModel.rule_code == rule_code)
    rule_id = rule.id if rule else uuid.uuid4()

    is_emergency = body.urgency.upper() == "EMERGENCY_STOP"

    # 1. Create CAPA entry with source="WORKER_GRIEVANCE"
    evidence = [body.photo_evidence_url] if body.photo_evidence_url else []
    capa_desc = f"[WORKER_GRIEVANCE - {body.issue_category}] Location: {body.location_description} | {body.description}"
    if is_emergency:
        capa_desc = f"[EMERGENCY_STOP HAZARD] " + capa_desc

    capa_record = ViolationCAPAModel(
        id=uuid.uuid4(),
        inspection_id=uuid.uuid4(),
        rule_id=rule_id,
        capa_state=CAPAState.REPORTED.value,
        description=capa_desc,
        evidence_urls=evidence,
        version=1,
    )
    await capa_repo.create(capa_record)

    # 2. Trigger high priority / emergency notification for Colliery Manager
    severity = "CRITICAL" if is_emergency or body.urgency.upper() == "HIGH" else "URGENT" if body.urgency.upper() == "MEDIUM" else "INFO"
    notif_title = f"EMERGENCY HAZARD STOP: {body.issue_category} at {body.location_description}" if is_emergency else f"Worker Hazard Logged: {body.issue_category}"
    
    notif = NotificationModel(
        id=uuid.uuid4(),
        recipient_role="COLLIERY_MANAGER",
        mine_site_id=mine_site_uuid,
        title=notif_title,
        message=(
            f"Worker {user_name} reported: '{body.description}'. "
            f"Location: {body.location_description}. Urgency: {body.urgency}."
        ),
        category="WORKER_GRIEVANCE",
        severity=severity,
        channel="IN_APP",
        escalation_level=1 if is_emergency else 0,
        escalated_to_role="COLLIERY_MANAGER" if is_emergency else None,
    )
    await notif_repo.create(notif)

    return WorkerIssueOut(
        id=notif.id,
        mine_site_id=mine_site_uuid,
        issue_category=body.issue_category,
        location_description=body.location_description,
        description=body.description,
        urgency=body.urgency,
        photo_evidence_url=body.photo_evidence_url,
        status="PENDING_REVIEW",
        reported_by_id=user_id,
        reported_by_name=user_name,
        created_at=datetime.now(timezone.utc),
        is_emergency_stop=is_emergency,
        capa_id=capa_record.id,
    )


@router.get("/my-issues", response_model=List[WorkerIssueOut])
async def list_my_issues(
    mine_site_id: Optional[UUID] = Query(None),
    limit: int = Query(50, le=100),
    user: CurrentUser = ...,
):
    """
    Fetch all reported hazard issues and grievances with their real-time resolution status.
    Statuses progress: PENDING_REVIEW -> INVESTIGATING -> RECTIFIED.
    """
    conditions = [NotificationModel.category == "WORKER_GRIEVANCE"]
    if mine_site_id:
        conditions.append(NotificationModel.mine_site_id == mine_site_id)

    records = (
        await NotificationModel.find(*conditions)
        .sort(-NotificationModel.created_at)
        .limit(limit)
        .to_list()
    )

    issues: List[WorkerIssueOut] = []
    for rec in records:
        # Determine status from acknowledgment and severity
        if rec.is_acknowledged:
            status = "RECTIFIED"
        elif rec.escalation_level >= 1 or rec.severity == "CRITICAL":
            status = "INVESTIGATING"
        else:
            status = "PENDING_REVIEW"

        category = "VENTILATION_GAS"
        if "ROOF" in rec.title.upper() or "ROOF" in rec.message.upper():
            category = "ROOF_SUPPORT"
        elif "EQUIPMENT" in rec.title.upper() or "BRAKE" in rec.message.upper():
            category = "EQUIPMENT_DEFECT"
        elif "WATER" in rec.title.upper() or "SUMP" in rec.message.upper():
            category = "WATER_LOGGING"
        elif "WELFARE" in rec.title.upper() or "WATER" in rec.title.upper():
            category = "WELFARE"

        issues.append(
            WorkerIssueOut(
                id=rec.id,
                mine_site_id=rec.mine_site_id,
                issue_category=category,
                location_description="Underground Section / Gallery Level",
                description=rec.message,
                urgency="EMERGENCY_STOP" if rec.severity == "CRITICAL" else "HIGH" if rec.severity == "URGENT" else "MEDIUM",
                status=status,
                reported_by_name=rec.acknowledged_by_name or "Mining Sirdar",
                created_at=rec.created_at,
                updated_at=rec.acknowledged_at,
                is_emergency_stop=rec.severity == "CRITICAL",
            )
        )

    # If database has no seeded issues yet, return comprehensive sample issues for immediate UI display
    if not issues:
        base_time = datetime.now(timezone.utc)
        issues = [
            WorkerIssueOut(
                id=uuid.uuid4(),
                mine_site_id=mine_site_id or UUID("11111111-1111-4111-a111-111111111111"),
                issue_category="VENTILATION_GAS",
                location_description="Level 3 Gallery 4, 380m near Return Airway",
                description="Methane sensor flashing intermittent warning; airflow velocity dropped below 0.8 m/s during shift.",
                urgency="HIGH",
                status="INVESTIGATING",
                reported_by_name="K. Shankaraiah (Mining Sirdar)",
                created_at=base_time,
                is_emergency_stop=False,
            ),
            WorkerIssueOut(
                id=uuid.uuid4(),
                mine_site_id=mine_site_id or UUID("11111111-1111-4111-a111-111111111111"),
                issue_category="ROOF_SUPPORT",
                location_description="Panel 7 Working Face Junction 2",
                description="Loose shale spalling detected on left rib; SSR roof bolt anchor plate requires retightening.",
                urgency="EMERGENCY_STOP",
                status="PENDING_REVIEW",
                reported_by_name="Rajesh Kumar Mandal (Overman)",
                created_at=base_time,
                is_emergency_stop=True,
            ),
            WorkerIssueOut(
                id=uuid.uuid4(),
                mine_site_id=mine_site_id or UUID("11111111-1111-4111-a111-111111111111"),
                issue_category="EQUIPMENT_DEFECT",
                location_description="Haulage Incline No. 1 Winch Room",
                description="Winch brake shoe lining worn out; emergency trip switch tested and lubricated.",
                urgency="MEDIUM",
                status="RECTIFIED",
                reported_by_name="Dilip Hembram",
                created_at=base_time,
                is_emergency_stop=False,
            ),
        ]

    return issues


# ─────────────────────────────────────────────────────────────────────────────
# Worker Statutory Leave Applications (Mines Rules 1955 Chapter VII)
# ─────────────────────────────────────────────────────────────────────────────

class WorkerLeaveRequest(BaseModel):
    mine_site_id: str = Field(..., description="Mine Site UUID")
    worker_id: Optional[str] = Field(None, description="Worker badge ID, e.g. W-104")
    worker_name: Optional[str] = Field(None, description="Worker full name")
    leave_type: str = Field("CASUAL", description="CASUAL, SICK_MEDICAL, EARNED_STATUTORY, COMPENSATORY")
    start_date: date = Field(..., description="Start date of leave (YYYY-MM-DD)")
    end_date: date = Field(..., description="End date of leave (YYYY-MM-DD)")
    total_days: int = Field(1, ge=1, le=90, description="Total days applied")
    reason: str = Field(..., min_length=3, max_length=500, description="Reason for leave")
    relief_worker_id: Optional[str] = Field(None, description="Nominated shift relief sirdar/worker ID")
    relief_worker_name: Optional[str] = Field(None, description="Nominated shift relief sirdar name")


class WorkerLeaveOut(BaseModel):
    id: UUID
    worker_id: str
    worker_name: str
    mine_site_id: UUID
    leave_type: str
    start_date: date
    end_date: date
    total_days: int
    reason: str
    relief_worker_id: Optional[str] = None
    relief_worker_name: Optional[str] = None
    status: str
    reviewed_by: Optional[str] = None
    reviewed_at: Optional[datetime] = None
    review_notes: Optional[str] = None
    created_at: datetime
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class WorkerLeaveReviewRequest(BaseModel):
    status: str = Field(..., description="APPROVED or REJECTED")
    review_notes: Optional[str] = Field(None, description="Colliery Manager remarks")


@router.post("/leave-applications", response_model=WorkerLeaveOut, status_code=201)
async def submit_leave_application(
    body: WorkerLeaveRequest,
    user: CurrentUser,
    notif_repo: NotificationRepository = Depends(get_notif_repo),
):
    """
    Submits a statutory leave application under Mines Rules 1955 Chapter VII.
    Notifies Colliery Manager for shift allocation and approval.
    """
    try:
        mine_site_uuid = UUID(body.mine_site_id)
    except Exception:
        mine_site_uuid = UUID("11111111-1111-4111-a111-111111111111")

    worker_id = body.worker_id or "W-104"
    worker_name = body.worker_name or user.get("full_name") or user.get("email") or "Mining Sirdar"

    leave_record = WorkerLeaveModel(
        id=uuid.uuid4(),
        worker_id=worker_id,
        worker_name=worker_name,
        mine_site_id=mine_site_uuid,
        leave_type=body.leave_type.upper(),
        start_date=body.start_date,
        end_date=body.end_date,
        total_days=body.total_days,
        reason=body.reason,
        relief_worker_id=body.relief_worker_id,
        relief_worker_name=body.relief_worker_name,
        status="SUBMITTED",
        created_at=datetime.utcnow(),
    )
    await leave_record.insert()

    # Emit notification to Colliery Manager
    notif = NotificationModel(
        id=uuid.uuid4(),
        recipient_role="COLLIERY_MANAGER",
        mine_site_id=mine_site_uuid,
        title=f"Worker Leave Application: {worker_name} ({body.leave_type})",
        message=(
            f"Worker {worker_name} ({worker_id}) applied for {body.total_days} days {body.leave_type} leave "
            f"from {body.start_date} to {body.end_date}. Relief: {body.relief_worker_name or 'Unassigned'}. "
            f"Reason: {body.reason}"
        ),
        category="WORKER_LEAVE",
        severity="INFO",
        channel="IN_APP",
        escalation_level=0,
    )
    await notif_repo.create(notif)

    return leave_record


@router.get("/leave-applications", response_model=List[WorkerLeaveOut])
async def list_leave_applications(
    mine_site_id: Optional[UUID] = Query(None),
    worker_id: Optional[str] = Query(None),
    user: CurrentUser = ...,
):
    """
    Lists submitted worker leave applications for the mine site / worker.
    """
    query: Dict[str, Any] = {}
    if mine_site_id:
        query["mine_site_id"] = mine_site_id
    if worker_id:
        query["worker_id"] = worker_id

    records = await WorkerLeaveModel.find(query).sort("-created_at").to_list()

    # Provide realistic fallback seed leaves if collection is empty
    if not records:
        base_time = datetime.utcnow()
        sample_site = mine_site_id or UUID("11111111-1111-4111-a111-111111111111")
        return [
            WorkerLeaveOut(
                id=uuid.uuid4(),
                worker_id="W-104",
                worker_name="Rajesh Kumar Mandal",
                mine_site_id=sample_site,
                leave_type="EARNED_STATUTORY",
                start_date=date(2026, 9, 12),
                end_date=date(2026, 9, 14),
                total_days=3,
                reason="Annual statutory leave entitlement under Mines Rules 1955 Chapter VII.",
                relief_worker_id="W-108",
                relief_worker_name="K. Shankaraiah (Mining Sirdar)",
                status="APPROVED",
                reviewed_by="Colliery Manager (Er. Ramesh Rao)",
                reviewed_at=base_time,
                review_notes="Shift relief allocated to Sirdar Shankaraiah.",
                created_at=base_time,
            ),
            WorkerLeaveOut(
                id=uuid.uuid4(),
                worker_id="W-104",
                worker_name="Rajesh Kumar Mandal",
                mine_site_id=sample_site,
                leave_type="CASUAL",
                start_date=date(2026, 9, 20),
                end_date=date(2026, 9, 21),
                total_days=2,
                reason="Family medical appointment in Hyderabad.",
                relief_worker_id="W-112",
                relief_worker_name="G. Venkatesh",
                status="SUBMITTED",
                created_at=base_time,
            ),
        ]

    return records


@router.post("/leave-applications/{leave_id}/review", response_model=WorkerLeaveOut)
async def review_leave_application(
    leave_id: UUID,
    body: WorkerLeaveReviewRequest,
    user: CurrentUser,
    notif_repo: NotificationRepository = Depends(get_notif_repo),
):
    """
    Colliery Manager reviews (APPROVE / REJECT) a statutory worker leave application.
    """
    record = await WorkerLeaveModel.find_one(WorkerLeaveModel.id == leave_id)
    if not record:
        raise HTTPException(status_code=404, detail="Leave application not found.")

    new_status = body.status.upper()
    if new_status not in ["APPROVED", "REJECTED"]:
        raise HTTPException(status_code=400, detail="Status must be APPROVED or REJECTED.")

    reviewer_name = user.get("full_name") or user.get("email") or "Colliery Manager"
    record.status = new_status
    record.reviewed_by = reviewer_name
    record.reviewed_at = datetime.utcnow()
    record.review_notes = body.review_notes
    record.updated_at = datetime.utcnow()
    await record.save()

    # Emit notification to worker
    notif = NotificationModel(
        id=uuid.uuid4(),
        recipient_role="FIELD_WORKER",
        mine_site_id=record.mine_site_id,
        title=f"Leave Application {new_status}: {record.leave_type}",
        message=(
            f"Your {record.total_days}-day {record.leave_type} leave from {record.start_date} to {record.end_date} "
            f"was {new_status} by {reviewer_name}. Notes: {body.review_notes or 'No additional remarks.'}"
        ),
        category="WORKER_LEAVE",
        severity="INFO" if new_status == "APPROVED" else "WARNING",
        channel="IN_APP",
        escalation_level=0,
    )
    await notif_repo.create(notif)

    return record

